### PR TITLE
SONARJAVA-1460 - Nullability constraint on an object checked with Objects null checking methods

### PR DESCRIPTION
**Pull request by DidierBesset**
…ects null checking methods

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NullCheckSymbolicValue.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ObjectConstraint;
+import org.sonar.java.se.ProgramState;
+import org.sonar.java.se.symbolicvalues.SymbolicValue.UnarySymbolicValue;
+
+import java.util.List;
+
+public class NullCheckSymbolicValue extends UnarySymbolicValue {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NullCheckSymbolicValue.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "checkPreStatement" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/cc01e73f2f804d0293f21a70d62897c722c1a87a/java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java#L66) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "checkPreStatement" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/cc01e73f2f804d0293f21a70d62897c722c1a87a/java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java#L66) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 2]
```diff
@@ -70,6 +73,14 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
     if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
       MethodInvocationTree methodInvocation = (MethodInvocationTree) syntaxNode;
       toCheck = methodInvocation.methodSelect();
+      if (ExplodedGraphWalker.isObjectsMethod(methodInvocation, REQUIRE_NON_NULL_METHOD_NAME, 1)) {
+        final List<SymbolicValue> values = context.getState().peekValues(2);
+        return context.getState().addConstraint(values.get(1), ObjectConstraint.NOT_NULL);
+      } else if (ExplodedGraphWalker.isObjectsMethod(methodInvocation, REQUIRE_NON_NULL_METHOD_NAME, 2)
+        || ExplodedGraphWalker.isObjectsMethod(methodInvocation, REQUIRE_NON_NULL_METHOD_NAME, 2)) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: CRITICAL Identical sub-expressions on both sides of operator "||" [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1764)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "checkPreStatement" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/eb5a4bb9941e710ee92187a8e9cb8421eea512ad/java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java#L66) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/test/files/se/NullDereferenceCheck.java [hunk 3]
```diff
@@ -1,7 +1,10 @@
 @SomeAnnotation(name = value)
 package javax.annotation;
 
+import java.util.function.Supplier;
```
Comment (m-g-sonar) [inline java-squid/src/test/files/se/NullDereferenceCheck.java]: these imports are useless as the code checking Objects is now in another file.

File: java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NullCheckSymbolicValue.java (hunk 1 above)
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NullCheckSymbolicValue.java]: should be an inner class of the null check, as it is only used there.

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 4]
```diff
@@ -574,6 +575,17 @@ private static boolean isLocalMethodInvocation(MethodInvocationTree tree) {
     return false;
   }
 
+  public static boolean isObjectsMethod(MethodInvocationTree syntaxNode, String methodName, int numberOfParameter) {
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: To me it has nothing to do in the ExplodedGraphWalker and should be moved in the ConstraintManager.

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 5]
```diff
@@ -81,6 +81,7 @@
   private static final int MAX_STEPS = 10000;
   private static final Logger LOG = LoggerFactory.getLogger(ExplodedGraphWalker.class);
   private static final Set<String> THIS_SUPER = ImmutableSet.of("this", "super");
+  public static final String OBJECT_CLASS_NAME = "java.lang.Object";
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: Never used, anywhere.

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 6]
```diff
@@ -574,6 +575,17 @@ private static boolean isLocalMethodInvocation(MethodInvocationTree tree) {
     return false;
   }
 
+  public static boolean isObjectsMethod(MethodInvocationTree syntaxNode, String methodName, int numberOfParameter) {
+    Symbol symbol = syntaxNode.symbol();
+    if (symbol.isMethodSymbol()) {
+      Symbol.MethodSymbol methodSymbol = (Symbol.MethodSymbol) symbol;
+      if (methodSymbol.owner().type().is("java.util.Objects") && methodName.equals(methodSymbol.name())) {
+        return methodSymbol.parameterTypes().size() == numberOfParameter;
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: handling of number of parameters is useless at this point and can be handled separately.

File: java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 7]
```diff
@@ -70,6 +73,14 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
     if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
       MethodInvocationTree methodInvocation = (MethodInvocationTree) syntaxNode;
       toCheck = methodInvocation.methodSelect();
+      if (ExplodedGraphWalker.isObjectsMethod(methodInvocation, REQUIRE_NON_NULL_METHOD_NAME, 1)) {
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: both if statement can be unified by looking a the number of parameters

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**